lib/dal/src/component/suggestion.rs

```diff
@@ -0,0 +1,686 @@
+/// Discovery as we will autosubscribe multiple components in that case
+/// When we need longer lived/more shared caching, we can move this elsewhere
+#[derive(Debug, Clone, Default)]
+pub struct PropSuggestionsCache {
+    /// Map of schema variant ID to its suggestion patterns
+    pub schema_suggestions: DashMap<SchemaVariantId, SchemaSuggestionMap>,
+    /// Map of schema name to components for efficient lookup
+    pub schema_to_components: DashMap<String, Vec<ComponentId>>,
+}
+
+/// Cache of prop suggestion patterns for a specific schema variant
+#[derive(Debug, Clone)]
+pub struct SchemaSuggestionMap {
+    /// Map of prop path -> list of prop suggestions for that path (suggestSources)
+    pub suggest_sources_map: HashMap<String, Vec<PropSuggestion>>,
+    /// Map of (target_schema, target_prop) -> list of source prop paths (suggestAsSourceFor)
+    pub suggest_as_source_for_map: HashMap<PropSuggestion, Vec<String>>,
+}
+
+impl PropSuggestionsCache {
+    /// Populate this cache instance with data from the database
+    /// Not sure how noisy this will be but I'd like to get the timings in the short term
+    #[instrument(level = "info", name = "dal.prop_suggestion_cache.populate", skip_all, 
+    fields(
+            si.change_set.id = Empty,
+            si.workspace.id = Empty,
+        ))]
+    pub async fn populate(&self, ctx: &DalContext) -> PropSuggestionCacheResult<()> {
+        let span = Span::current();
+        span.record("si.change_set.id", ctx.change_set_id().to_string());
+        span.record(
+            "si.workspace.id",
+            ctx.tenancy()
+                .workspace_pk_opt()
+                .unwrap_or(WorkspacePk::NONE)
+                .to_string(),
+        );
+        // Build schema_to_components map
+        for component_id in Component::list_ids(ctx).await? {
+            let comp_schema = Component::schema_for_component_id(ctx, component_id).await?;
+            let schema_name = comp_schema.name().to_string();
+
+            // Update or insert the component list for this schema
+            self.schema_to_components
+                .entry(schema_name)
+                .or_default()
+                .push(component_id);
+        }
+
+        // Build schema suggestions map for each unique schema variant
+        let mut processed_schema_variants = HashSet::new();
+
+        for component_id in Component::list_ids(ctx).await? {
+            let schema_variant_id = Component::schema_variant_id(ctx, component_id).await?;
+
+            if processed_schema_variants.contains(&schema_variant_id) {
+                continue; // Already processed this schema variant
+            }
+            processed_schema_variants.insert(schema_variant_id);
+
+            // Build suggestion map for this schema variant
+            let mut suggest_sources_map = HashMap::new();
+            let mut suggest_as_source_for_map = HashMap::new();
```

Comment:
Can we unify these two maps so we only have to do one pass when finding matches.

That is, if you detect a suggestAsSourceFor, transform that into a suggestSources?